Polishing changes (#87)
canonical/ubuntu-maintainers-handbook: PackageBuilding.md

to a PPA will be built with the same machinery as official Ubuntu packages, so
they're also a great way to verify your work before formally submitting it to
Ubuntu.

Once built, the packages in a PPA are publically available, allowing you to
share them with bug reporters for testing fixes, and to reviewers of your merge
proposals who can readily use them for testing.

The downsides to PPAs are:

1. They share resources with other Launchpad build processes, so during busy
   times it can take a while to come up in the build queue.
2. It's less hands-on than a local build so can be hard to use for highly
   iterative workflows like sorting out dependency issues or git-bisecting
   build failures (if you expect that you'll need to debug the build, like
   going into the environment to modify and retry, a local build is
   recommended - see below).
3. The PPAs are picky about version strings.

### Set the version string

For the PPA, we need to change the version in the changelog that's lower than
the official version we plan to release. Since the tilde `~` character sorts
lower than everything else in launchpad, we can simply append `~<string>1` to

> the version string in `debian/changelog`. For example:

Author: I found the wording in this paragraph really confusing, not sure I can really make any suggestions as I'm not sure what's being conveyed

> For the PPA, we need to change the version in the changelog that's lower than
> the official version we plan to release. Since the tilde `~` character sorts
> lower than everything else in launchpad, we can simply append `~<string>1` to
> the version string in `debian/changelog`. For example:
> 
> ```bash
> -postfix (3.3.0-1ubuntu0.1) bionic; urgency=medium
> +postfix (3.3.0-1ubuntu0.1~bionic1) bionic; urgency=medium
> ```

Reviewer: > I found the wording in this paragraph really confusing, not sure I can really make any suggestions as I'm not sure what's being conveyed

@s-makin What do you think about this?

````suggestion
For the PPA, we should change the version in the changelog to one that 
is lower than the official version we plan to release. For example:

```bash
-postfix (3.3.0-1ubuntu0.1) bionic; urgency=medium
+postfix (3.3.0-1ubuntu0.1~bionic1) bionic; urgency=medium
```

Since the tilde `~` character sorts lower than everything else in launchpad, 
we can simply append `~<string>1` to the version string in `debian/changelog`. 
(See more details about the sorting algorithm here: [`deb-version(7)`](https://manpages.ubuntu.com/manpages/lunar/en/man7/deb-version.7.html))
````

Author: Such a small change, but I think it's much clearer now - thanks!

https://launchpad.net/~kstenerud/+archive/ubuntu/postfix-postconf-segfault-1753470/+packages

> **Note**:
> You must wait for the package to build server-side before you can use the
> PPA to install packages. This might take anywhere from a few minutes to a
> few hours depending on how busy things are!
> 
> It'll first build the binaries for each architecture, then publish the
> source and binary packages to be publically downloadable.

#### Check progress with `ppa`

You can use the `ppa` tool to poll launchpad for progress status:

```bash
$ ppa wait ppa:kstenerud/postfix-sru-lp1753470-segfault
```

It will exit with `0` once the PPA packages have fully built.

Launchpad also sends "status updates" notification mails, so monitor your
inbox.

## Build binary packages locally with `sbuild

Assuming you have configured `sbuild` properly, you can use it to build the
binary package:

```bash
$ sbuild
```

Because of https://bugs.launchpad.net/launchpad/+bug/1699763, it is a good
idea to disable the inclusion of `.buildinfo` files in the `*_source.changes`
 file:

```bash
$ sbuild --debbuildopts='--buildinfo-option=-O'
```

For more information, see:

> * https://packaging.ubuntu.com/html/setting-up-sbuild.html

Reviewer: The Ubuntu-Maintainers Guide has a setup section for `sbuild` itself. See https://github.com/canonical/ubuntu-maintainers-handbook/blob/main/Setup.md#software-sbuild
